core/src/test/groovy/io/micronaut/core/annotation/AnnotationValueSpec.groovy

```diff
@@ -0,0 +1,122 @@
+package io.micronaut.core.annotation
+
+import spock.lang.Specification
+
+class AnnotationValueSpec extends Specification {
+
+    void "test class value"() {
+        given:
+        def av = AnnotationValue.builder("test.Foo")
+                       .value(AnnotationValueSpec)
+                       .build()
+
+        expect:
+        av.classValue().get() == AnnotationValueSpec
+        av.classValue("value").get() == AnnotationValueSpec
+        av.classValue("value", Specification).get() == AnnotationValueSpec
+        !av.classValue("value", URL).isPresent()
+    }
+
+    void "test class value 2"() {
+        given:
+        def av = AnnotationValue.builder("test.Foo")
+                .values(new AnnotationClassValue<Object>(AnnotationValueSpec), new AnnotationClassValue<Object>(Specification))
+                .build()
+
+        expect:
+        av.classValues().length == 2
+        av.classValues()[0] == AnnotationValueSpec
+        av.classValues()[1] == Specification
+        av.classValue().get() == AnnotationValueSpec
+        av.classValue("value").get() == AnnotationValueSpec
+        av.classValue("value", Specification).get() == AnnotationValueSpec
+        !av.classValue("value", URL).isPresent()
+    }
+
+    void "test INT value"() {
+        given:
+        def av = AnnotationValue.builder("test.Foo")
+                .value(10)
+                .build()
+
+        expect:
+        av.intValue().asInt == 10
+    }
+
+    void "test LONG value"() {
+        given:
+        def av = AnnotationValue.builder("test.Foo")
+                .value(10)
+                .member("str", "10")
+                .build()
+
+        expect:
+        av.longValue().asLong == 10
+        av.longValue("str").asLong == 10
+    }
+
+    void "test string value"() {
+        given:
+        def av = AnnotationValue.builder("test.Foo")
+                .member("number", 10)
+                .member("bool", true)
+                .member("type", new AnnotationClassValue(Specification))
+                .member("types", new AnnotationClassValue(Specification), new AnnotationClassValue(AnnotationValueSpec))
+                .build()
+
+        expect:
+        av.stringValue("number").get() == "10"
+        av.stringValue("bool").get() == "true"
+        av.stringValue("type").get() == Specification.name
+        av.stringValue("types").get() == Specification.name
+        av.stringValues("number") == ["10"] as String[]
+    }
+
+
+    void "test INT value array"() {
+        given:
+        int[] ints = [10, 20]
+        def av = AnnotationValue.builder("test.Foo")
+                .values(ints)
+                .build()
+
+        expect:
+        av.intValue().asInt == 10
```

Comment:
I think this behavior is somewhat surprising. I can imagine someone using this and only receiving the first value and assuming that was all that was set in the annotation. Perhaps we should throw an exception in this case

Reply:
Annotation values are bit different, it is common to use arrays to represent "no value" in annotation definitions even if you only care about the first value. eg. `Class[] foo() default {}` or `int[] bar() default {}`

This is just a convenience method so that libraries can specify if they only care about a single `int` value or many. What is probably missing from the impl is all the variants for array types (like `intValues()` etc.)